### ISSUE TITLE
Feature request: nix repl history navigation via prefix string like fish shell

### ISSUE THREAD

**Issue by lolbinarycat** [NixOS/nix#10167]
**Is your feature request related to a problem? Please describe.**
often when in `nix repl` you want to re-evaluate an expression.  currently you have 3 options:
* press up dozens of times until you reach the expression you want
* retype the expression in full
* use ^R (clunky, and has poor discoverability)

**Describe the solution you'd like**
the fish shell (and several other REPLs) have a solution for this:
when pressing up with a string already entered into the prompt, instead of just going back 1 step in history, it goes back until it finds a command starting with that string.

**Describe alternatives you've considered**
some way to set aliases for frequently used commands via something like `:alias`.

**Additional context**
bringing nixpkgs (or whatever nix expression i am working into scope of the repl often feels tedious, expecially with rapid iterative development.

`:r` should help with this slightly, but not as much if you have to leave the repl to do a git command or something.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by l0b0** [comment]
This can be configured for readline in general by adding the following content to `~/.inputrc`:

```
"\e[A": history-search-backward
"\e[B": history-search-forward
```

It [looks like the Nix REPL uses something called `editline`](https://github.com/NixOS/nix/blob/1907a3300fd6a36cea6bee3e19acc707053a01d3/src/libcmd/package.nix#L30) rather than `readline` on Linux/Mac. ~~Maybe that can be configured the same way.~~ Unfortunately [that can't be configured this way](https://discourse.nixos.org/t/does-nix-repl-use-editrc-on-linux/69319/4).

Maybe we should change the issue to ask for the _much_ more common and flexible `readline` instead of `editline`? (At least on Linux, in case `readline` isn't well supported on Mac.)

**Comment by xokdvium** [comment]
Nix be built with either: https://github.com/NixOS/nix/blob/master/src/libcmd/meson.build#L48C1-L57C6

**Comment by l0b0** [comment]
> Nix be built with either: [`master`/src/libcmd/meson.build#L48C1-L57C6](https://github.com/NixOS/nix/blob/master/src/libcmd/meson.build#L48C1-L57C6)

Yep, so it should hopefully be easy (as these things go, when there are thousands of users) to change.

To be clear, I'm hoping that unless there are reasons beyond closure size to leave this as is, we could move to readline as the default. I expect a lot of people would rather not build their own custom Nix.
